hpx/runtime/serialization/detail/polymorphic_nonintrusive_factory.hpp

```diff
@@ -0,0 +1,166 @@
+//  Copyright (c) 2014 Thomas Heller
+//  Copyright (c) 2015 Anton Bikineev
+//
+//  Distributed under the Boost Software License, Version 1.0.
+//  See accompanying file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_SERIALIZATION_POLYMORPHIC_NONINTRUSIVE_FACTORY_HPP
+#define HPX_SERIALIZATION_POLYMORPHIC_NONINTRUSIVE_FACTORY_HPP
+
+#include <hpx/runtime/serialization/serialization_fwd.hpp>
+#include <hpx/util/jenkins_hash.hpp>
+#include <hpx/util/static.hpp>
+#include <hpx/util/demangle_helper.hpp>
+#include <hpx/traits/polymorphic_traits.hpp>
+
+#include <boost/noncopyable.hpp>
+#include <boost/unordered_map.hpp>
+#include <boost/mpl/bool.hpp>
+#include <boost/type_traits/is_abstract.hpp>
+
+namespace hpx { namespace serialization { namespace detail
+{
+    struct function_bunch_type
+    {
+        typedef void (*save_function_type) (output_archive& , const void* base);
+        typedef void (*load_function_type) (input_archive& , void* base);
+        typedef void* (*create_function_type) ();
+
+        save_function_type save_function;
+        load_function_type load_function;
+        create_function_type create_function;
+    };
+
+    class HPX_EXPORT polymorphic_nonintrusive_factory: boost::noncopyable
+    {
+    public:
+        typedef boost::unordered_map<std::string,
+                  function_bunch_type, hpx::util::jenkins_hash> serializer_map_type;
+
+        static polymorphic_nonintrusive_factory& instance()
+        {
+            hpx::util::static_<polymorphic_nonintrusive_factory> factory;
+            return factory.get();
+        }
+
+        void register_class(const std::string& class_name,
+            const function_bunch_type& bunch)
+        {
+             map_[class_name] = bunch;
```

Comment:
This could probably benefit from one of the `emplace` or `insert` overloads, and also clarify the intent on replacement.